### DISCUSSION TITLE
Use custom configuration

### DISCUSSION THREAD

**Discussion by oosavu** [rust-skia/rust-skia#1069] [Q&A]
Hi!
I am trying to use this crate on web/emscripten as a part of my c++ project. 
Specifically the  call stack should be like this: `JavaScript -> my_old_c++_code -> my_rust_code -> skia_safe -> c++_skia_lib`.
For this reason I should to have versatile way to configure and build SKIA (e.g. set specific skia_system_freetype2_lib path, many skia_use_* flags...)
It is possible to set a`gn` arguments for build SKIA itself manually? 

As i understood, skia_bindings crate provide several abilities to build skia manually.  One of them is "adjusting BuildConfiguration and FinalBuildConfiguration" in `skia-buindings/build.rs` file.
Well.. How to do it? :) 
Do i have to clone this repo and modify bild.rs file? or use cargo patches feature?
The most convenient way for me would be to do it just using command-line arguments. Is it possible?
Or may be i can just build skia by myself and then just set SKIA_LIBRARY_SEARCH_PATH while skia_safe will be build?

**Comment by pragmatrix** [comment]
> It is possible to set a gn arguments for build SKIA itself manually?

Only if you build Skia yourself. We could add support to pull additional arguments from an environment variable.

> One of them is "adjusting BuildConfiguration and FinalBuildConfiguration" in skia-buindings/build.rs file.
Well.. How to do it? :)

These are just internal types and not meant to be extended (yet).

> Do i have to clone this repo and modify build.rs file?

I am not entirely sure, but this surely be a start for experimentation.

> or use cargo patches feature?

I am not sure how this could help.

> The most convenient way for me would be to do it just using command-line arguments. Is it possible?

If you mean building Skia by yourself, sure. The build process of skia-binding's build.rs is completely automatic and allows only for relative minor configuration via environment variables.

> Or may be i can just build skia by myself and then just set SKIA_LIBRARY_SEARCH_PATH while skia_safe will be build?

`SKIA_LIBRARY_SEARCH_PATH` together with (`SKIA_SOURCE_DIR`](https://github.com/rust-skia/rust-skia/tree/master/skia-bindings#changing-the-skia-source-directory) could work here.

**Reply by oosavu** [reply]
Thanks for reply!
For now i am planning to use SKIA_LIBRARY_SEARCH_PATH / SKIA_SOURCE_DIR. But it is complex solution due to CI/CD reasons (i have to provide SKIA sources to the development environment for all building images and local PC and manage SKIA version). 

>We could add support to pull additional arguments from an environment variable.

It will be great! For my opinion the clear solution would be:
1) Add support for custom command-line argument. e.g. SKIA_CUSTOM_GN_ARGS
2) Parse this argument in the build.rs, and then force rewrite FinalBuildConfiguration.gn_args  just before SKIA configuration command called.

with this approach we will provide defaults for the most of users, but people who needs versatile customisation will be able to do it at their own risk.

Do you accept contribution? I can try to do it myself
